web-api/src/storage/elastic.rs

```diff
@@ -115,79 +116,38 @@ impl Client {
         }
+        let mut knn_scores = self.search_request(&body).await?;
+
         if let Some(query) = params.query {
+            body.remove("knn");
             filter.insert("must".to_string(), json!({ "match": { "snippet": query }}));
             body.insert("query".to_string(), json!({ "bool": filter }));
-            body.insert("explain".to_string(), json!(true));
-        }
 
-        let scores = self
-            .query_with_json::<_, SearchResponse<NoSource>>(
-                self.create_resource_path(["_search"], None),
-                Some(body),
-            )
-            .await?
-            .map(|response| {
-                if params.query.is_some() {
-                    // mixed knn and bm25 scores need to be normalized
-                    const KNN_DESCRIPTION: &str = "within top k documents";
-                    let mut max_bm25_score = None;
-                    let scores = response
-                        .hits
-                        .hits
-                        .into_iter()
-                        .map(|hit| {
-                            // details has exactly one or two elements: knn or bm25 or both
-                            let mut knn_score = None;
-                            let mut bm25_score = None;
-                            let detail = &hit.explanation.details[0];
-                            if detail.description == KNN_DESCRIPTION {
-                                knn_score = Some(detail.value);
-                            } else {
-                                bm25_score = Some(detail.value);
-                            }
-                            if let Some(detail) = hit.explanation.details.get(1) {
-                                if detail.description == KNN_DESCRIPTION {
-                                    knn_score = Some(detail.value);
-                                } else {
-                                    bm25_score = Some(detail.value);
-                                }
-                            }
-                            if bm25_score > max_bm25_score {
-                                max_bm25_score = bm25_score;
-                            }
-                            (hit.id, knn_score, bm25_score)
-                        })
-                        .collect_vec();
+            let bm25_scores = self.search_request(body).await?;
+            let max_bm25_score = bm25_scores
+                .values()
+                .copied()
+                .max_by(nan_safe_f32_cmp)
+                .unwrap_or_default()
+                .max(1.0);
+
+            // mixed knn and bm25 scores need to be normalized
+            for (id, bm25_score) in bm25_scores {
+                let bm25_score = bm25_score / max_bm25_score;
+                knn_scores
+                    .entry(id)
+                    .and_modify(|knn_score| *knn_score = 0.5 * *knn_score + 0.5 * bm25_score)
+                    .or_insert(bm25_score);
+            }
```

Comment:
(but we do now run it in two requests, because there where issues when running it with one requests?)

Reply:
the post processing for the score normalization stays the same. the issue was with the returned results from es. let me try to give an example. lets say es does a hybrid search and finds those docs with scores
```
knn         bm25
"1": 0.6    "4": 10
"2": 0.5    "5": 1
"3": 0.4    "6": 0.1
```
and we request a count of 3 then we will get the docs `"4", "5", "1"` because es knows nothing about our normalized scores and on those three we would do our postprocessing. but what we actually want is that es uses the normalized scores to pick the top 3, which would be
```
knn         bm25
"1": 0.6    "4": 1
"2": 0.5    "5": 0.1
"3": 0.4    "6": 0.01
```
and we would get `"4", "1", "2"`. since we can't tell the es scripts to do the normalization for us (the script only has access to the current hit, but it can't compute max/sum/avg/etc) we need to do two separate queries and do the merging and post-processing in our own. the example would be more complicated if knn and bm25 return overlapping doc ids, but the idea stays the same.